### ISSUE TITLE
Slow networking between two containers (tried host and bridge)

### ISSUE THREAD

**Issue by esabol** [docker/docker#30801]
I have two containers on the same host (RHEL 7.2, 3.10.0-514.6.1.el7.x86_64 SMP, dm-thinpool storage) that I'm networking together. Network transfers between the two containers take twice as long compared to a setup with the client code running natively outside of Docker on the same host.

Container A is a server which listens on port 47301. I start it like this:

`docker run -d --restart=unless-stopped -p 47301:47301 -p 9000:9000 -v /somedir:/somedir:ro --name container-a container-a:latest`

Container B connects to the server in Container A **using SSL**. I start it like this:

`docker run -d  --restart=unless-stopped --network=host --name container-b container-b:latest`

The code in Container B is set to connect to myhostname.fqdn:47301.

The "host" network shows container B is connected (although, curiously, I don't see any mention of container A):
```
> docker network inspect host
[
    {
        "Name": "host",
        "Id": "adablahblah",
        "Created": "2017-02-01T09:18:21.348731359-05:00",
        "Scope": "local",
        "Driver": "host",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": null,
            "Config": []
        },
        "Internal": false,
        "Attachable": false,
        "Containers": {
            "743blahblahblah": {
                "Name": "container-b",
                "EndpointID": "af6yadayadayada",
                "MacAddress": "",
                "IPv4Address": "",
                "IPv6Address": ""
            }
        },
        "Options": {},
        "Labels": {}
    }
]
```

**I also tried creating my own bridge network:**

```
> docker network create --driver bridge myproject
```
Then, I specified `--network=myproject` when starting both container A and container B. In that scenario, the code in container-B is set to connect to `container-b:47301`. This works fine, but there is no improvement in performance. The network transfers still take about twice as long when compared to the container-B code running natively on the same host (outside of Docker).

My Docker daemon does not have IPv6 enabled.

**Output of `docker version`:**

```
Client:
 Version:      1.13.0
 API version:  1.25
 Go version:   go1.7.3
 Git commit:   49bf474
 Built:        Tue Jan 17 09:55:28 2017
 OS/Arch:      linux/amd64

Server:
 Version:      1.13.0
 API version:  1.25 (minimum version 1.12)
 Go version:   go1.7.3
 Git commit:   49bf474
 Built:        Tue Jan 17 09:55:28 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 3
 Running: 3
 Paused: 0
 Stopped: 0
Images: 181
Server Version: 1.13.0
Storage Driver: devicemapper
 Pool Name: os-dkthinpool
 Pool Blocksize: 524.3 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file: 
 Metadata file: 
 Data Space Used: 15.35 GB
 Data Space Total: 51.64 GB
 Data Space Available: 36.29 GB
 Metadata Space Used: 4.583 MB
 Metadata Space Total: 25.17 MB
 Metadata Space Available: 20.58 MB
 Thin Pool Minimum Free Space: 5.163 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: true
 Deferred Deletion Enabled: true
 Deferred Deleted Device Count: 0
 Library Version: 1.02.135-RHEL7 (2016-11-16)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 03e5862ec0d8d3b3f750e19fca3ee367e13c090e
runc version: 2f7393a47307a16f8cee44a37b262e8b81021e3e
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-514.6.1.el7.x86_64
Operating System: Scientific Linux 7.2 (Nitrogen)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 3.686 GiB
Name: myhostname
ID: stuff
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

Can anyone help? I'd expect a ~10% decrease in networking performance using bridge/bridge container networking. A ~100% decrease in performance is unacceptable for our project.

**Comment by cpuguy83** [comment]
From what I can tell from your configuration here, you are using an exposed port on the host to communicate back into the container.
This is most definitely going to be much slower because all local traffic must go through a proxy process, this is due to limitations in linux bridge networking.
One way to get around this and continue using it as you are is to set `--userland-proxy=false` on the daemon which enables hairpin bridge mode... the reason this is not the default is because of known kernel issues which cause major system degradation....
I would not recommend doing this.

Really I would have the containers communicate directly rather than going through this port forward.
The easiest way to do this is to create a network:

```
$ docker network create myApp
```

And then have each container use this network:

```
$ docker run --net=myApp <other options>
```

Then each container can communicate with the other via container name.

**Comment by esabol** [comment]
Thanks for the reply, @cpuguy83, but I tried creating a network for the containers to communicate directly. I mentioned this in my initial post. Please search for "I also tried creating my own bridge network". This did not improve performance though. Did I create the network incorrectly? I used `docker network create --driver bridge myproject`.

**Comment by esabol** [comment]
Here's some additional information on the network I created:

```
> docker network inspect myproject
[
    {
        "Name": "myproject",
        "Id": "359742ea0b1df97aad4b068f352e9863ed6f80f27a9fe3bfbb37808ecb57af87",
        "Created": "2017-02-06T19:51:55.877408686-05:00",
        "Scope": "local",
        "Driver": "bridge",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": {},
            "Config": [
                {
                    "Subnet": "172.18.0.0/16",
                    "Gateway": "172.18.0.1"
                }
            ]
        },
        "Internal": false,
        "Attachable": false,
        "Containers": {
            "37d9aablahblahblahblahblahblahblahblahblahblah": {
                "Name": "container-a",
                "EndpointID": "ddd4f19380f626c37e233972da77111d1d50a274d2c0c0d0eafecb80fe609e15",
                "MacAddress": "02:42:xx:xx:00:02",
                "IPv4Address": "172.18.0.2/16",
                "IPv6Address": ""
            },
            "b7197blahblahblahblahblahblahblahblahblahblah": {
                "Name": "container-b",
                "EndpointID": "b6c62f785882a4b627d9d3c32417d858341fe13a45927689eff96dcef8ef9b3d",
                "MacAddress": "02:42:xx:xx:00:03",
                "IPv4Address": "172.18.0.3/16",
                "IPv6Address": ""
            }
        },
        "Options": {},
        "Labels": {}
    }
]
```

As you can see, it shows both containers are connected to the `myproject` network that I created. The code in container B is set to connect to "container-a:47301". My benchmarks actually shows this setup to be slightly slower than if I have container-b start with `--network=host`. Am I doing something wrong?

**Comment by cpuguy83** [comment]
ping @aboch

**Comment by aboch** [comment]
@esabol  besides what @cpuguy83 already pointed out, I do not have an idea where this 100% penalty may come from. 

How do you measure the network performance, what do you look at ?

Others have done some throughput measurement (https://www.percona.com/blog/2016/08/03/testing-docker-multi-host-network-performance/) and they saw 12% decrease for containers on the bridge network compared to the apps running outside of containers.

**Comment by esabol** [comment]
Thanks for the reply, @aboch. I'm familiar with that percona.com web page, which is why I expected ~12%, but I got ~100%. I would be happy with ~12%, but I cannot reproduce their results. That article also doesn't document their Docker configuration or how they created their networks, so it's particularly helpful.

I benchmark 30 iterations, and I'm looking wallclock times. Each iteration takes ~1 second on average on the host OS, but ~2 seconds in Docker, regardless of `--network=myproject` or `--network=host`. I have tried running the benchmark code inside container B, and I've also run the benchmark code outside of Docker on the host OS. I get the same results.

Does my bridge network seem correct to you?

I saw an article on stackoverflow.com that suggested enabling ipv6 in the Docker daemon might improve networking performance, but they were talking about a much older version of Docker. Do you have an opinion on that?

Are there any other settings in daemon.json that you think might be applicable (besides `userland-proxy`)?

**Comment by esabol** [comment]
I continue to bang my head on this. I got the idea to try moving the two systems into the same container to see whether or not it really was inter-container networking which was slowing me down or something else. I saw no substantial improvement in networking performance. I saw the same roughly 100% slower networking than the code executing natively on the host system. In this scenario, the server was listening on port 47301 as before. I tried it with and without `--network=host`, with `--network=myproject`, and with and without exposing ports with `-p`.

I'm starting to wonder if this is some issue specifically with RHEL/CentOS 7.2 and/or the network devices on the host.

My next course of action is try to get the Docker daemon running with ipv6 enabled to see if that makes a difference.

If anyone has any other advice or suggestions as to what I can do to resolve this issue, please let me know.

**Comment by aboch** [comment]
What are you measuring ? Big file transfer ? When you run your your test natively on the host, are both server and client process running on the same host ? If so consider the MTU constraints when they run inside containers.

**Comment by esabol** [comment]
They're not big transfers, no. Probably about 300-500 bytes.

Yes, when I test natively on the host, the server and client processes are running on the same host. Same scenario as having both run inside the same container, basically. The tasks are run serially, not in parallel.

My host OS's MTU is 1500. That's what I see inside the Docker container as well.

Host networking devices:
````
> ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: ens192: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP qlen 1000
    link/ether 00:50:56:96:62:cc brd ff:ff:ff:ff:ff:ff
    inet 128.183.16.36/23 brd 128.183.17.255 scope global ens192
       valid_lft forever preferred_lft forever
    inet6 fe80::250:56ff:fe96:62cc/64 scope link 
       valid_lft forever preferred_lft forever
3: docker0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP 
    link/ether 02:42:17:58:e3:d3 brd ff:ff:ff:ff:ff:ff
    inet 172.17.0.1/16 scope global docker0
       valid_lft forever preferred_lft forever
    inet6 fe80::42:17ff:fe58:e3d3/64 scope link 
       valid_lft forever preferred_lft forever
4: virbr0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN qlen 1000
    link/ether 52:54:00:ed:c0:d2 brd ff:ff:ff:ff:ff:ff
    inet 192.168.122.1/24 brd 192.168.122.255 scope global virbr0
       valid_lft forever preferred_lft forever
5: virbr0-nic: <BROADCAST,MULTICAST> mtu 1500 qdisc pfifo_fast master virbr0 state DOWN qlen 1000
    link/ether 52:54:00:ed:c0:d2 brd ff:ff:ff:ff:ff:ff
391: br-359742ea0b1d: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN 
    link/ether 02:42:3e:1d:41:c1 brd ff:ff:ff:ff:ff:ff
    inet 172.18.0.1/16 scope global br-359742ea0b1d
       valid_lft forever preferred_lft forever
    inet6 fe80::42:3eff:fe1d:41c1/64 scope link 
       valid_lft forever preferred_lft forever
411: veth9a03940@if410: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master docker0 state UP 
    link/ether 42:xx:e1:68:61:84 brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet6 fe80::40d8:e1ff:fe68:6184/64 scope link 
       valid_lft forever preferred_lft forever
````

Inside of Docker container:

````
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: ens192: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 00:50:56:96:62:cc brd ff:ff:ff:ff:ff:ff
    inet 128.183.16.36/23 brd 128.183.17.255 scope global ens192
       valid_lft forever preferred_lft forever
    inet6 fe80::250:56ff:fe96:62cc/64 scope link 
       valid_lft forever preferred_lft forever
3: docker0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default 
    link/ether 02:42:17:58:e3:d3 brd ff:ff:ff:ff:ff:ff
    inet 172.17.0.1/16 scope global docker0
       valid_lft forever preferred_lft forever
    inet6 fe80::42:17ff:fe58:e3d3/64 scope link 
       valid_lft forever preferred_lft forever
4: virbr0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN group default qlen 1000
    link/ether 52:54:00:ed:c0:d2 brd ff:ff:ff:ff:ff:ff
    inet 192.168.122.1/24 brd 192.168.122.255 scope global virbr0
       valid_lft forever preferred_lft forever
5: virbr0-nic: <BROADCAST,MULTICAST> mtu 1500 qdisc pfifo_fast master virbr0 state DOWN group default qlen 1000
    link/ether 52:54:00:ed:c0:d2 brd ff:ff:ff:ff:ff:ff
391: br-359742ea0b1d: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN group default 
    link/ether 02:42:3e:1d:41:c1 brd ff:ff:ff:ff:ff:ff
    inet 172.18.0.1/16 scope global br-359742ea0b1d
       valid_lft forever preferred_lft forever
    inet6 fe80::42:3eff:fe1d:41c1/64 scope link 
       valid_lft forever preferred_lft forever
411: veth9a03940@if410: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master docker0 state UP group default 
    link/ether 42:d8:e1:68:61:84 brd ff:ff:ff:ff:ff:ff
    inet6 fe80::40d8:e1ff:fe68:6184/64 scope link 
       valid_lft forever preferred_lft forever
````

**Comment by esabol** [comment]
And here's inside the Docker container with `--network=myproject`:
````
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
554: eth0@if555: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default 
    link/ether 02:42:ac:12:00:02 brd ff:ff:ff:ff:ff:ff
    inet 172.18.0.2/16 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::42:acff:fe12:2/64 scope link 
       valid_lft forever preferred_lft forever
````

**Comment by esabol** [comment]
Enabling IPv6 didn't help. In fact, it appears to have made the networking performance worse in my benchmarks. We also upgraded from Docker 1.13.0 to 1.13.1 at the same time (probably shouldn't have conflated the two changes, but oh well), but the net result was 20-30% worse. :(

This is what we did:
````
$ ip -6 route add 2001:db8:1::/80 dev docker0
$ sysctl net.ipv6.conf.default.forwarding=1
$ sysctl net.ipv6.conf.all.forwarding=1
````
Then, added
````
  "ipv6": true,
  "fixed-cidr-v6": "2001:dba:1::/80",
````
to our daemon.json file. Most of that comes straight from the https://docs.docker.com/engine/userguide/networking/default_network/ipv6/ but we went with "80" instead of "64" for the subnet.

I benchmarked with `--network=host` and with `--network=myproject6` in which the myproject6 network was created like so:
````
$ docker network create myproject6 --driver bridge --ipv6 --subnet "2001:db8:1::/80".
````

We undid all that since it didn't improve anything. We're probably going to try upgrading from CentOS 7.2 to 7.3 on the host next, but I'm not holding out much hope for that improving performance.

**Comment by esabol** [comment]
Upgrading to CentOS 7.3 didn't help, of course.

Going to try `userland-proxy: "false"` next, even though I have been advised above that it's problematic. Pointers to more information on that would be useful. I haven't found much discussion of it.

I wonder if why I'm seeing so much slower networking performance using docker-proxy is because my host only has 4 cores? Is there an optimal number of CPU cores for docker-proxy usage?

**Comment by esabol** [comment]
`userland-proxy: "false"` didn't seem to help either! Something seems seriously amiss. Found the following lines in /var/log/messages, which might explain it:

````
Feb 17 13:56:47 myhostname kernel: docker0: port 2(veth76e42d9) entered disabled state
Feb 17 13:44:38 myhostname audispd: node=heabox type=SYSCALL msg=audit(1487357078.174:4156): arch=c000003e syscall=54 success=yes exit=0 a0=4 a1=0 a2=40 a3=129be20 items=0 ppid=1021 pid=24517 auid=4294967295 uid=0 gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=(none) ses=4294967295 comm="iptables" exe="/usr/sbin/xtables-multi" subj=system_u:system_r:iptables_t:s0 key=(null)
Feb 17 13:56:47 myhostname firewalld: WARNING: COMMAND_FAILED: '/usr/sbin/iptables -w2 -t nat -C DOCKER -p tcp -d 0/0 --dport 47301 -j DNAT --to-destination 172.17.
0.3:47301' failed:
Feb 17 13:56:47 myhostname firewalld: WARNING: COMMAND_FAILED: '/usr/sbin/iptables -w2 -t nat -C POSTROUTING -p tcp -s 172.17.0.3 -d 172.17.0.3 --dport 47301 -j MASQUERADE' failed:
````

Can anyone spare a clue? Thanks!

**Comment by esabol** [comment]
The iptables commands shown in the previous comment should execute properly, right?

**Comment by BastianHofmann** [comment]
@esabol Did you find the issue? I think I'm hitting the same problem.

**Comment by esabol** [comment]
It's good to get to hear that someone else is experiencing the same problem, @BastianHofmann. No, I've found no solution, and the Docker devs stopped responding to my posts here. :(

With nothing to go on, I gave up on running our client in a container. The performance penalty wasn't worth it. We still run the server in a container, but the rest was wasted effort.

I suppose I should test it again after the next Docker upgrade to make sure the situation hasn't improved. I suspect the issue is with RHEL/CentOS 7.x. Perhaps a newer kernel doesn't have this problem?

**Comment by BastianHofmann** [comment]
@esabol the issue came up during research I'm doing for a thesis.

My best guess is that the kernel version or distro is at fault. In my setup I used the buildpack-deps:jessie image. I can provide you with more detailed information about my setup later today.

Just to confirm with you: The issue you are seeing includes slow network performance of a process running inside a container with --net=host (i.e. the fact that the process is running inside the container causes the network slow down, even with --net=host)?

**Comment by cpuguy83** [comment]
If you are using `--net=host`, there is no network virtualisation happening you are directly on the host network stack.
So it sounds like if you are seeing slowness there's something else at play here.

**Comment by elsauto** [comment]
Hi folks, I just found this thread. The very moment I switched to bridge (external) networking I can hardly get a query response from my MySQL server (in example). Are there any work arounds this? Any help would be much apreciated. See you all around.

**Comment by esabol** [comment]
@BastianHofmann, I tried it ever possible way I could: a separate named bridge network, `--net=host`, `userland-proxy: false`, moving everything inside a single container, IPv4, and IPv6, and I encountered poor networking performance with all of them when compared to running the client code natively. I ultimately decided it was a kernel/distro issue with either CentOS 7.3 (my host OS) or Debian Jessie (the distro in the containers). Have you tried different distros? I should retest with the latest Docker version.

@elsauto, I never found any workaround (other than not using a container for my client code at all), but you might want to try starting your containers with `--net=host` and/or putting `userland-proxy: false` in your docker configuration. These things are supposed to help, but they didn't help in my tests. Good luck!

**Comment by cpuguy83** [comment]
Are we sure it's the network throughput that is the issue and not the application?
`--net=host` literally runs the application on the host networking stack, there is no bridging, nating, or anything involved.

**Comment by BastianHofmann** [comment]
I resolved my issue. The performance hit only happens after I install the Docker daemon. First I compared the performance of the application running in the Docker container with the application running natively, but without Docker installed. There are some Docker background jobs running, but I'm not sure what they do or how to disabled them (flame graph showed something with health checks).

**Comment by esabol** [comment]
> Are we sure it's the network throughput that is the issue and not the application?

If/when I retest, I'll strip the code bare and test only the networking code, but I'll point you back to my post in this issue from February 7th in which I wrote: "I added timings to the work being done, and it's constant, regardless of whether it is done inside of a container or not, so it seems like the only difference is networking...."

The data being sent over the network per SSL connection is relatively small, ~800 bytes, give or take 200 bytes. My benchmarks tested between a hundred and a thousand of such connections, and the code running inside of the container was consistently about 100% slower (in wallclock timings) than running it natively.

**Comment by maksymilian-palka** [comment]
Hi,

I have encountered that performance issue too :(
I've tried both bride and --network=host options.
Latter increases performance, but still it does not reach the performance of app outside of the docker.

My tests are a little bit different: I have used two web frameworks: C++ Wt webtoolkit and Iris.
Iris seems to not have so big degradation:
* bridge: from 130000 reqs/s (no docker) drops to 99000 reqs/s
C++ Wt has very big degradation:
* bridge: from 114000 reqs/s (no docker) drops to 55000 reqs/s !!

I need to repeat those tests with '--network=host' on the original setup, but when tried on virtual machines, then I have following:

**HOST (no docker)**
c++: around 4.1 MB/s - ~29700 reqs/sec
Iris: around 5.3 MB/s - ~38000 reqs/sec

**Bridge**
c++: around 2.7 MB/s - ~19950 reqs/sec
Iris: around 4.4 MB/s - ~31600 reqs/sec

**--network=host**
c++: around 2.9 MB/s - ~20900 reqs/sec
Iris: around 5.1 MB/s - ~36000 reqs/sec

Again - Iris seems to be more stable (--network=host) while C++ Wt suffers when running in Docker.

**Comment by BastianHofmann** [comment]
@maksymilian-palka I forgot to mention this here, but I also saw a increase in throughput performance when using the --privileged flag on container creation. Apparently accessing host devices (e.g. the network interfaces for receiving and sending packets) is slowed down by some kind of capability check. The --privileged flag should remove those capability checks.
I plan to release a paper containing more information about this issue soon. I will link it when it's done.
In the mean time can you try to confirm my results by using the privileged flag in your setup?

**Comment by BastianHofmann** [comment]
Could also be apparmor checks. If you search for "Privileged" in https://github.com/moby/moby/blob/master/daemon/oci_linux.go you can see all the checks which are disabled when turn on privileged (I think docs only mentions capability flags and devices, but there are more).

**Comment by maksymilian-palka** [comment]
@BastianHofmann I have run both dockers with --privilaged:
**--network=host --privileged**
c++: around 3.38 MB/s - ~24000
Iris: around 5.29 MB/s - ~38000

I think it is not the network issue anymore, because Iris has almost the same results inside as on outside of the docker. Any idea what could influence an execution of C++ code which depends on few libraries (libwt, libhttp, boost)? Both docker images are alpine based and WT lib was compiled within a docker.

**Comment by BastianHofmann** [comment]
@maksymilian-palka maybe file access (e.g. for logging)? The container is using some storage driver (don't know what you have configured) and the container file system is usually slower than the hosts (you could use a volume to mount the relevant part of the host system). Other than that I'm not sure what it could be.

**Comment by BastianHofmann** [comment]
@maksymilian-palka please note that I do not recommend the usage of --privileged in a production environment. You should identify what capability checks cause the slow down and allow only what is needed by the container.

**Comment by maksymilian-palka** [comment]
@BastianHofmann, thank you! I will check if there is something more happening when C++ code is being executed. I will write when I will find something.

**Comment by maksymilian-palka** [comment]
@BastianHofmann, your comment about logging gave me a hint. Normally WT lib is logging to stdout/stderr. I was forwarding it to /dev/null. And this was the source of the difference. When I have recompiled WT lib and removed all logging, then the results were close to each other (inside/outside of the docker). When added --privileged flag, then the results were similar.
Thank you!

**Comment by BastianHofmann** [comment]
@esabol Did that solve your issue as well? Maybe we can close the issue then since the performance hit is kind of expected.

**Comment by esabol** [comment]
We just upgraded to CentOS 7.4 and Docker 17.09.0-ce this week. I'll see if I can re-test soon.

**Comment by BastianHofmann** [comment]
Could you also try the option: “--security-opt apparmor=unconfined” instead of privileged (with net host)? I suspect app armor is causing the drop in performance. Privileged is setting this option but also does other things.

**Comment by esabol** [comment]
No. I've been meaning to go back and re-test/re-benchmark everything after having upgraded a bunch of things, but haven't gotten around to it, unfortunately.

My solution was not to run my worker/client inside a Docker container (container B above), which isn't a solution, but it resulted in the networking performance we needed.

**Comment by frostybeard** [comment]
We work with the Android AOSP, and doing a "repo sync -j4" (4 threads) inside a docker container was taking upwards of 24 hours, where the same "repo sync -j4" operation on the same build server outside the container took around 2 hours.  So I guess we were seeing this problem. I am not a docker expert so if you want debugging information, tell me what do do.

**Comment by xaduha** [comment]
My issue was with headless chrome and was solved by `--disable-dev-shm-usage` flag. Try running in privileged mode or passing `/dev` directly, see if there is any difference.

**Comment by polarathene** [comment]
Sharing for others benefit. I used `docker network create` and overhead of docker networking was not that bad (Ubuntu 20.10, kernel 5.8, Docker 19.03), but when adding `--subnet` option to the network, with no other changes network performance suffered greatly. 9 minutes vs 2 hours for some network testing task (testssl.sh).
